ksql-cli/src/test/java/io/confluent/ksql/cli/SslClientAuthFunctionalTest.java

```diff
@@ -0,0 +1,289 @@
+    sslContextFactory = new Server();
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseCliIfClientDoesNotProvideCertificate() {
+
+    // Given:
+    givenClientConfiguredWithoutCertificate();
+
+    // Then:
+    expectedException.expect(KsqlRestClientException.class);
+    expectedException.expectCause(is(instanceOf(ExecutionException.class)));
+    expectedException.expectCause(hasCause(is(instanceOf(SSLHandshakeException.class))));
+
+    // When:
+    canMakeCliRequest();
+  }
+
+  @Test
+  public void shouldBeAbleToUseCliOverHttps() {
+    // Given:
+    givenClientConfiguredWithCertificate();
+
+    // When:
+    final Code result = canMakeCliRequest();
+
+    // Then:
+    assertThat(result, is(Code.OK));
+  }
+
+  @Test
+  public void shouldNotBeAbleToUseWssIfClientDoesNotTrustServerCert() throws Exception {
+    // Given:
+    givenClientConfiguredWithoutCertificate();
+
+    // Then:
+    expectedException.expect(SSLHandshakeException.class);
+    expectedException.expectMessage("Received fatal alert: handshake_failure");
+
+    // When:
+    makeWsRequest();
+  }
+
+  @Test
+  public void shouldBeAbleToUseWss() throws Exception {
+    // Given:
+    givenClientConfiguredWithCertificate();
+
+    // When:
+    makeWsRequest();
+
+    // Then: did not throw.
+  }
+
+  private void givenClientConfiguredWithCertificate() {
+
+    String clientCertPath = ServerKeyStore.keyStoreProps()
+        .get(SslConfigs.SSL_TRUSTSTORE_LOCATION_CONFIG);
+    String clientCertPassword = ServerKeyStore.keyStoreProps()
+        .get(SslConfigs.SSL_TRUSTSTORE_PASSWORD_CONFIG);
+
+    // HTTP:
+    clientProps = new HashMap<>();
+    clientProps.putAll(ClientTrustStore.trustStoreProps());
+
+    clientProps.put(SslConfigs.SSL_KEYSTORE_LOCATION_CONFIG, clientCertPath);
+    clientProps.put(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG, clientCertPassword);
+
+    clientProps.put(KsqlClient.DISABLE_HOSTNAME_VERIFICATION_PROP_NAME, "true");
+    clientProps.put(KsqlClient.TLS_ENABLED_PROP_NAME, "true");
+
+    // WS:
+    sslContextFactory.setTrustStorePath(ClientTrustStore.trustStorePath());
+    sslContextFactory.setTrustStorePassword(ClientTrustStore.trustStorePassword());
+    sslContextFactory.setKeyStorePath(clientCertPath);
+    sslContextFactory.setKeyStorePassword(clientCertPassword);
+    sslContextFactory.setEndpointIdentificationAlgorithm("");
+  }
+
+  private void givenClientConfiguredWithoutCertificate() {
+
+    // HTTP:
+    clientProps = new HashMap<>();
+    clientProps.putAll(ClientTrustStore.trustStoreProps());
+
+    clientProps.put(KsqlClient.DISABLE_HOSTNAME_VERIFICATION_PROP_NAME, "true");
+    clientProps.put(KsqlClient.TLS_ENABLED_PROP_NAME, "true");
+
+    // WS:
+    sslContextFactory.setTrustStorePath(ClientTrustStore.trustStorePath());
+    sslContextFactory.setTrustStorePassword(ClientTrustStore.trustStorePassword());
+    sslContextFactory.setEndpointIdentificationAlgorithm("");
+  }
+
+  private Code canMakeCliRequest() {
```

Comment:
nit: de-dup these helper methods between this file and `SslFunctionalTest.java`?

Reply:
I tried to do this originally. But unfortunately SslFunctionalTest uses static test fixtures so I can't push things up into a common base class as I would like to :(

Reply:
Not sure I follow. Can we not de-dup the methods by making them static, e.g., 
```
private Code canMakeCliRequest() { ... }
```
becomes
```
static Code canMakeCliRequest(final TestKsqlRestApp restApp, final Map<String, String> clientProps) { ... }
```
(and similarly for the others)?

Reply:
Then to keep the individual tests cleaner each can have:
```
private Code canMakeCliRequest() {
    return canMakeCliRequest(REST_APP, clientProps);
}
```
(and similar).

These methods would still have to be duplicated between the two tests but they seem tiny enough and the actual implementation of the helper methods would only appear in one place. WDYT?